### PR TITLE
publishing docker images to ghcr.io on release

### LINKED ISSUE #85: Publish docker images to ghcr.io
## Summary
With each release, we should build a docker image and publish it to the [CDCgov packages repo](https://github.com/orgs/CDCgov/packages) with the appropriate tags.

## Acceptance Criteria
- [ ] A new CI workflow that auto publishes an image
- [ ] Another condition that defines task completion.

## Details / Tasks
The workflow should only kick off after a new final release (not draft release) has been created in the repo.

## Dependencies
#64

### PR DESCRIPTION
**Pull request by ericbuckley**
## Description
On release, new Docker images will be pushed to the [ghcr.io package repo](https://github.com/CDCgov/RecordLinker/pkgs/container/recordlinker)

## Related Issues
closes #85 

## Additional Notes
- upgrading all the CI checkout actions to v4 because of a deprecation warning
- small updates to the Dockerfile to be more oriented towards production and moving the development code to overrides in the compose.yml
- removing otel traces from link.py

<--------------------- REMOVE THE LINES BELOW BEFORE MERGING --------------------->

## Checklist
Please review and complete the following checklist before submitting your pull request:

- [x] I have ensured that the pull request is of a manageable size, allowing it to be reviewed within a single session.
- [x] I have reviewed my changes to ensure they are clear, concise, and well-documented.
- [x] I have updated the documentation, if applicable.
- [x] I have added or updated test cases to cover my changes, if applicable.
- [x] I have minimized the number of reviewers to include only those essential for the review.

## Checklist for Reviewers
Please review and complete the following checklist during the review process:

- [ ] The code follows best practices and conventions.
- [ ] The changes implement the desired functionality or fix the reported issue.
- [ ] The tests cover the new changes and pass successfully.
- [ ] Any potential edge cases or error scenarios have been considered.

### CODE REVIEW

File: src/recordlinker/linking/link.py [hunk 1]
```diff
@@ -136,48 +133,43 @@ def link_record_against_mpi(
     # find the highest scoring match across all passes
     scores: dict[models.Person, float] = collections.defaultdict(float)
     for algorithm_pass in algorithm.passes:
-        with TRACER.start_as_current_span("link.pass"):
```
[the remark this answers is not in the document]
Comment (ericbuckley) [reply src/recordlinker/linking/link.py:139]: To reduce the requirements necessary to build the app.  The OTEL sdk and api libraries add a bunch of dependencies, and this really isn't necessary for production use, as these traces were only being evaluated in the performance testing suite.

We could create a mock tracker for this though and leave them in.  I didn't have that thought until just now.  What do you think about that?

File: compose.yml [hunk 2]
```diff
@@ -27,6 +27,8 @@ services:
     build:
       context: .
       args:
+        ENVIRONMENT: dev
```
Comment (alhayward) [inline compose.yml:30]: Why are some environment defaults `dev` and some `prod`? (I imagine this is a software engineering design practice, would love to learn more)

File: src/recordlinker/linking/link.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (alhayward) [reply src/recordlinker/linking/link.py:139]: Do we anticipate wanting to evaluate the performance testing suite in production? If so, then IIUC maybe worth keeping in the mock.

**Review by alhayward — APPROVED** [review]
Less of my wheelhouse, so it was a great opportunity to learn more! Left 2 questions.

File: src/recordlinker/linking/link.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ericbuckley) [reply src/recordlinker/linking/link.py]: I added the spans back and used a mock tracer for the cases in production where otel is not installed.

https://github.com/CDCgov/RecordLinker/pull/94/commits/ca6c2ab9461d5874df662f2f27d7028934253fef

File: compose.yml (hunk 2 above)
Comment (ericbuckley) [reply compose.yml:30]: The `Dockerfile` will be used in production environments.  In some upcoming work, I have a release process that publishes these containers to our ghcr.io package repo.  From there NBS, and hopefully other future STLTs, will download the image and install it in their environments.  From a customer's perspective, defaulting to "prod" makes a lot of sense.

The compose.yml is just something used by developers for testing and getting things working locally.  It's a convenience layer to get things up and running, but never used in production environments.  Theoretically ops teams could use something like a compose.yml to setup cloud and on-prem environments, but I've never seen that done before, everyone these days leans into [K8s](https://kubernetes.io/) or vendor specific solutions like [ECS](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/Welcome.html).

**Review by cbrinson-rise8 — APPROVED** [review]

**Pull request merged**